docs/proposals/all-in-one-volume.md

```diff
@@ -0,0 +1,333 @@
+        - path: "labels"
+          fieldRef:
+            fieldPath: metadata.labels
+        - path: "annotations"
+          fieldRef:
+            fieldPath: metadata.annotations
+  - name: config-volume
+    configMap:
+      name: special-config
+      items:
+      - key: special.how
+        path: path/to/special-key
+```
+
+## Analysis
+
+There are several combinations of resources that can be used at once, which
+all warrant consideration. The combinations are listed with one instance of
+each resource, but real world usage will support multiple instances of a
+specific resource too.
+
+### ConfigMap + Secrets + Downward API
+
+The user wishes to deploy containers with configuration data that includes
+passwords. An application using these resources could be deploying OpenStack
+on Kubernetes. The configuration data may need to be assembled differently
+depending on if the services are going to be used for production or for
+testing. If a pod is labeled with production or testing, the downward API
+selector metadata.labels can be used to produce the correct OpenStack configs.
+
+### ConfigMap + Secrets
+
+Again, the user wishes to deploy containers with configuration data that
+includes passwords. In this case with MariaDB running, the operator may wish
+the container to have a ~/.my.cnf file that includes the username and password
+for the database.
+
+### ConfigMap + Downward API
+
+In this case, the user wishes to generate a config including the pod’s name
+(available via the metadata.name selector). This application may then pass the
+pod name along with requests in order to easily determine the source without
+using IP tracking.
+
+### Secrets + Downward API
+
+A user may wish to use a secret as a public key to encrypt the namespace of
+the pod (available via the metadata.namespace selector). This example may be
+the most contrived, but perhaps the operator wishes to use the application to
+deliver the namespace information securely without using an encrypted
+transport.
+
+### Collisions between keys when all keys from a resource are projected
+
+In the event of a user not explicitly defining keys for pathing within a volume,
+there can not be an overlap of resources with the same paths on a given volume.
+If a conflict does occur, a numeric prefix will be put in the name signifying
+the order that the resources were specified in the pod spec. An example of this
+resolution is as follows:
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: volume-test
+spec:
+  containers:
+  - name: container-test
+    image: busybox
+    volumeMounts:
+    - name: all-in-one
+      mountPath: "/system-volume"
+      readOnly: true
+  volumes:
+  - name: all-in-one
+    system:
+      items:
+      - secretName: mysecret
+      - configMapName: myconfigmap
+```
+
+Here the assumption is that the data values configured for each resource is the
+same. The contents of /system-volume may look like this:
+
+/system-volume/1-very-generic
```

Comment:
I don't think a numeric prefix is especially usable.  How about:

1.  `<type of resource>-<resource name>-<key>`
2.  `<resource name>-<key>`